workflow/controller/dag.go

```diff
@@ -186,34 +186,29 @@ func (d *dagContext) hasMoreRetries(node *wfv1.NodeStatus) bool {
 		return true
 	}
 	// pick the first child to determine it's template type
-	childNode := d.wf.Status.Nodes[node.Children[0]]
-	tmpl, err := d.tmplCtx.GetTemplate(&childNode)
-	if err != nil {
+	childNode, ok := d.wf.Status.Nodes[node.Children[0]]
+	if !ok {
 		return false
 	}
-	if tmpl.RetryStrategy.Limit != nil && int32(len(node.Children)) > *tmpl.RetryStrategy.Limit {
+	tmpl := d.wf.GetStoredOrLocalTemplate(&childNode)
+	if tmpl != nil && tmpl.RetryStrategy != nil && tmpl.RetryStrategy.Limit != nil && int32(len(node.Children)) > *tmpl.RetryStrategy.Limit {
 		return false
 	}
 	return true
 }
 
-func (woc *wfOperationCtx) executeDAG(nodeName string, tmplCtx *templateresolution.Context, tmpl *wfv1.Template, orgTmpl wfv1.TemplateHolder, boundaryID string) *wfv1.NodeStatus {
-	node := woc.getNodeByName(nodeName)
-	if node != nil && node.Completed() {
-		return node
-	}
-	if node == nil {
-		node = woc.initializeNode(nodeName, wfv1.NodeTypeDAG, orgTmpl, boundaryID, wfv1.NodeRunning)
-	}
+func (woc *wfOperationCtx) executeDAG(nodeName string, tmplCtx *templateresolution.Context, tmpl *wfv1.Template, boundaryID string) error {
+	node := woc.markNodePhase(nodeName, wfv1.NodeRunning)
```

Comment:
I see that the logic has been changed to unconditionally mark this node as Running. Previously, executeDAG was a NO-OP when the dag node was already considered completed.  The new logic means we will flap back to running no matter what and go through re-assessment logic.

I'd like to understand the reason for this change since it seems like we should never flap back to running.

Reply:
I changed the logic and node completions are checked at the beginning of `executeTemplate`. So, I assume a node is not completed in `executeDAG`.

https://github.com/argoproj/argo/pull/1552/files#diff-fcb04129f1d32e69cca32631c6586587R1108-R1114

Reply:
I know `executeDAG` can be called outside of `executeTemplate` currently and it makes a node status unexpected. We need a big refactoring to avoid this unexpected call by class structuring. I'm totally fine to put completion checks in sub-methods called from `executeTemplate` for now.